p2panda-net/src/lib.rs

```diff
@@ -1,5 +1,140 @@
+//! topologies on top of BLE (Bluetooth Low Energy), LoRa or even Digital Radio Communication
+//! infrastructure.
+//!
+//! ## Extensions
+//!
+//! `p2panda-net` is agnostic to any data type (sending and receiving raw byte streams) and can
+//! seamlessly be extended with external or official p2panda implementations for different parts of
+//! the application:
+//!
+//! 1. Custom Data types exchanged over the network
+//! 2. Optional relay nodes to aid connection establishment when peers are behind firewalls etc.
+//! 3. Custom sync protocol for any data types, with managed re-attempts on connection failures and
+//!    optional re-sync schedules
+//! 4. Custom peer discovery strategies (multiple approaches can be used at the same time)
+//! 5. Sync and storage of (very) large blobs
+//! 6. Fine-tune gossipping behaviour
+//! 7. Additional custom protocol handlers
+//!
+//! ## Integration with other p2panda solutions
+//!
+//! We provide p2panda's fork-tolerant and prunable append-only logs in `p2panda-core`, offering
+//! single-writer and multi-writer streams, authentication, deletion, ordering and more. This can
+//! be further extended with an efficient sync implementation in `p2panda-sync` and validation and
+//! fast stream-based ingest solutions in `p2panda-streams`.
+//!
+//! For discovery of peers on the local network, we provide an mDNS-based implementation in
+//! `p2panda-discovery`, planned next are additional techniques like "rendesvouz" nodes and random
+//! walk algorithms.
+//!
+//! Lastly we maintain persistance layer APIs in `p2panda-store` for in-memory storage or
+//! embeddable, SQL-based databases.
+//!
+//! In the future we will provide additional implementations for managing access control and group
+//! encryption.
+//!
+//! ## Example
+//!
+//! ```
+//! # use anyhow::Result;
+//! use p2panda_core::{PrivateKey, Hash};
+//! use p2panda_discovery::mdns::LocalDiscovery;
+//! use p2panda_net::{NetworkBuilder, TopicId};
+//! use p2panda_sync::Topic;
+//! use serde::{Serialize, Deserialize};
+//! # #[tokio::main]
+//! # async fn main() -> Result<()> {
+//!
+//! // Peers knowing the same "network id" will eventually find each other. This is the most global
+//! // identifier to group peers into multiple networks when necessary.
+//! let network_id = [1; 32];
+//!
+//! // The network can be used to automatically find and ask other peers about any data the
+//! // application is interested in. This is expressed through "network-wide queries" over topics.
+//! //
+//! // In this example we would like to be able to query messages from each chat group, identified
+//! // by a BLAKE3 hash.
+//! #[derive(Clone, Debug, PartialEq, Eq, Hash, Deserialize, Serialize)]
+//! struct ChatGroup(Hash);
+//!
+//! impl ChatGroup {
+//!     pub fn new(name: &str) -> Self {
+//!         Self(Hash::new(name.as_bytes()))
+//!     }
+//! }
+//!
+//! impl Topic for ChatGroup {}
```

Comment:
Seeing the example like this really makes me wish we'd renamed `Topic` to `Query` :rofl: